service/history/chasm_engine.go

```diff
@@ -0,0 +1,204 @@
+// The MIT License
+//
+// Copyright (c) 2025 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/chasm"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/locks"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/consts"
+	historyi "go.temporal.io/server/service/history/interfaces"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow/cache"
+)
+
+type ChasmEngine struct {
+	entityCache     cache.Cache
+	shardController shard.Controller
+	registry        *chasm.Registry
+	configs         *configs.Config
+}
+
+func NewChasmEngine(
+	entityCache cache.Cache,
+	shardController shard.Controller,
+	registry *chasm.Registry,
+	configs *configs.Config,
+) *ChasmEngine {
+	return &ChasmEngine{
+		entityCache:     entityCache,
+		shardController: shardController,
+		registry:        registry,
+		configs:         configs,
+	}
+}
+
+func (e *ChasmEngine) UpdateComponent(
+	ctx context.Context,
+	ref chasm.ComponentRef,
+	updateFn func(chasm.MutableContext, chasm.Component) error,
+	opts ...chasm.TransitionOption,
+) (updatedRef chasm.ComponentRef, retError error) {
+
+	shardContext, executionLease, err := e.getExecutionLease(ctx, ref)
+	if err != nil {
+		return chasm.ComponentRef{}, err
+	}
+	defer func() {
+		executionLease.GetReleaseFn()(retError)
+	}()
+
+	mutableState := executionLease.GetMutableState()
+	chasmTree, ok := mutableState.ChasmTree().(*chasm.Node)
+	if !ok {
+		return chasm.ComponentRef{}, serviceerror.NewInternal(
+			fmt.Sprintf("CHASM tree implementation not properly wired up, encountered type: %T", mutableState.ChasmTree()),
```

Comment:
Let's agree on `CHASM` vs `chasm` in error messages. I am fine with both, but `CHASM` looks better to me (it is an abbreviation after all).

Reply:
also
```suggestion
			fmt.Sprintf("CHASM tree implementation not properly wired up, encountered type: %T, expected type: %T", mutableState.ChasmTree(), &chasm.Node{}),
```

Reply:
+1 for `CHASM` as well.

Updated the message.